### PR TITLE
docs: Pin docker to version 18.06

### LINKED ISSUE #373: Decide on docker version strategy
Docker removed devicemapper in 18.09 which complicates life for many (see https://github.com/kata-containers/runtime/pull/1112).

Our [install docs](https://github.com/kata-containers/documentation/tree/master/install) setup the docker repo and then help users *install the latest version of docker*. Given the above, that's a potential problem so we need to find a non-gross way of updating the install docs to install docker 18.06.

The approach needs to:

- be simple.
- work with [`kata-manager`](https://github.com/kata-containers/tests/tree/master/cmd/kata-manager).

Since the docker package will get auto-upgraded from 18.06 -> 18.09 when the user upgrades their installed packages, we're basically going to have to `apt-mark hold` docker I think:

- https://github.com/kata-containers/runtime/issues/926#issuecomment-440689612

We'll also have to find the appropriate distro rune for each other install doc and then remove it if/when docker can be upgraded once again. And add a "note" explaining what the extra step is for.

/cc @egernst.

### PR DESCRIPTION
**Pull request by jodh-intel**
Docker 18.09 removed devicemapper support but did not provide an
alternative. This can cause problems for users so update the install
docs to install Docker at version 18.06 (the last version that supports
devicemapper).

This is a temporary solution until either docker provide an alternative
or we find a way to work around the Docker feature being removed.

Fixes #373.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
And now for the caveats...

- I cannot test RHEL or SLES so we'll need input from others on that
  (@GabyCT, @marcov, @flavio, @mssola).

- fwics, docker 18.06 is not / no longer available for Fedora 29:

  https://download.docker.com/linux/fedora/29/x86_64/stable/Packages

  I've added in a minimal fix to deal with that but the summary is: if you're running F29, you'll be running Docker 18.09 (meaning no devicemapper).

/cc @egernst.

**Comment by jodh-intel** [comment]
Some of the CI builds are failing with:

```sh
Cloning into 'tests'...
/home/centos/tests/.ci/lib.sh: line 19: GOPATH: unbound variable
```

**Comment by jodh-intel** [comment]
Fedora CI:

```sh
[kata-containers-documentation-fedora-PR] $ /bin/bash /tmp/jenkins7239932866990890670.sh
Cloning into 'tests'...
/home/fedora/tests/.ci/lib.sh: line 19: GOPATH: unbound variable
Build step 'Execute shell' marked build as failure
Setting status of 85dc9c7d17f69987e7512e9395e5ff7e4c8e4e06 to FAILURE with url http://jenkins.katacontainers.io/job/kata-containers-documentation-fedora-PR/50/ and message: 'Build finished. '
Using context: jenkins-ci-fedora
Finished: FAILURE
```

CentOS CI:

```sh
[kata-containers-documentation-centos-7-4-PR] $ /bin/bash /tmp/jenkins1058249980225164497.sh
Cloning into 'tests'...
/home/centos/tests/.ci/lib.sh: line 19: GOPATH: unbound variable
Build step 'Execute shell' marked build as failure
Setting status of 85dc9c7d17f69987e7512e9395e5ff7e4c8e4e06 to FAILURE with url http://jenkins.katacontainers.io/job/kata-containers-documentation-centos-7-4-PR/56/ and message: 'Build finished. '
Using context: jenkins-ci-centos-7-4
Finished: FAILURE
```

**Comment by jodh-intel** [comment]
Blocked on https://github.com/kata-containers/tests/pull/1134, which should fix the `GOPATH` error.

**Comment by jodh-intel** [comment]
This PR is effectively also blocked on https://github.com/kata-containers/tests/pull/1136 I think as we need to tweak the CI scripts to remove docker between running each documentation test.

File: install/docker/opensuse-docker-install.md [hunk 1]
```diff
@@ -5,13 +5,24 @@
...
+   >   devicemapper support.
+   > - To remove the lock on the docker package to allow it to be updated:
+   >   ```sh
+   >   $ sudo zypper removelock docker
+   >   ```
 
    ```bash
    $ sudo zypper -n install libcgroup1
-   $ sudo zypper -n install docker
+   $ sudo zypper -n install -f docker-18.06.1_ce-51.1
```
Comment (marcov) [inline install/docker/opensuse-docker-install.md]: Another neat way to do this without hard coding a specific docker version is:

`docker -n install docker<18.09'`

This is going to work on SLES too.

**Comment by GabyCT** [comment]
@jodh-intel tested on OpenSuse and it is working, I do not have a license anymore of RHEL.

**Comment by marcov** [comment]
@jodh-intel fyi the docker packages for SUSE/SLE are built with a custom commands, and the `devicemapper` storage driver is still available in the `18.09` version we distribute.

**Comment by chavafg** [comment]
https://github.com/kata-containers/tests/pull/1134 merged

Now I see this error:
```
1 upgraded, 1 newly installed, 0 to remove and 125 not upgraded.
E: Held packages were changed and -y was used without --allow-change-held-packages.
Build step 'Execute shell' marked build as failure
```

**Comment by jodh-intel** [comment]
@chavafg - yep, I need to update the doc CI tests on this PR to remove docker between running each doc test now that https://github.com/kata-containers/tests/pull/1136 is merged...

File: install/docker/opensuse-docker-install.md (hunk 1 above)
Comment (jodh-intel) [reply install/docker/opensuse-docker-install.md]: Thanks @marcov - I've updated the branch for opensuse + sles.

**Comment by jodh-intel** [comment]
@fatherlinux, @mrguitar, @lsm5 - could you provide any input from the RHEL perspective? We don't have licenses any more and therefore cannot test updates to the RHEL install docs here.

**Comment by jodh-intel** [comment]
Now blocked on https://github.com/kata-containers/tests/pull/1144.

**Comment by jodh-intel** [comment]
Huzzah! Atlast we have a CI pass for this PR! ;)

---

@chavafg, @GabyCT - btw, I spotted this in the Jenkins logs:

```
Switched to branch 'PR_374'
Current branch PR_374 is up to date.
bash: .ci/hypervisors//filter_docker_.sh: No such file or directory
make -C cmd/checkcommits
```

**Comment by grahamwhaley** [comment]
hmm, maybe I need to enable github ack checks on this repo..... can I get an ack from @marcov maybe before we actually press the button?
I'm going to DNM it until that point, but am v.happy if others agree that is over cautions and go ahead and merge....

**Comment by jodh-intel** [comment]
@marcov, @mssola, @flavio - could you comment on this?

@egernst - I know you were keen to see this land so could you also ptal?

**Comment by krsna1729** [comment]
Quick question @jodh-intel - Any reason why we don't do `curl -fsSL https://get.docker.com | VERSION=18.06 sh`, which supports several distros, other than friends don't let friends do `curl-pipe-bash` :P

**Comment by krsna1729** [comment]
also 18.06.2 vs 18.06.1 in the current changes to use the one patched for the runc CVE?
https://docs.docker.com/engine/release-notes/#18062

**Comment by jcvenegas** [comment]
@jodh-intel please rebase this PR.

**Comment by jodh-intel** [comment]
@jcvenegas - branch updated.

@krsna1729 - Updated for 18.06.2. Regarding your comment about using get.docker.com, that needs a lot more discussion and is not appropriate for this PR imho. You are of course free to install docker how you wish. However, although the get.docker.com approach is convenient, I'm not convinced many users would be happy running an opaque script downloaded off the web as `root`, particularly on a server. Yes, I know you can browse to that site and see the script, but it could change. Or the site could get hacked, or, or, or... By providing clear instructions in our installation guides, admins comfortable with shell script can see exactly what is happening.

If you wish to raise a PR using the get.docker.com approach, we can maybe force the issue and get more input from the rest of the community.

**Comment by jodh-intel** [comment]
Ping @kata-containers/documentation - The dnm was added by @grahamwhaley to get an extra review or two I think but this should have landed long ago. Please review.

File: install/docker/centos-docker-install.md [hunk 2]
```diff
@@ -5,13 +5,24 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/centos-docker-install.md]: Suggested rewrite: 
so the following commands install the latest version, which includes

File: install/docker/debian-docker-install.md [hunk 3]
```diff
@@ -6,16 +6,27 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/debian-docker-install.md]: Suggested rewrite:
so the following commands install the latest version, which includes

File: install/docker/fedora-docker-install.md [hunk 4]
```diff
@@ -5,14 +5,28 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/fedora-docker-install.md]: Suggested rewrite:
so the following commands install the latest version, which includes

File: install/docker/opensuse-docker-install.md [hunk 5]
```diff
@@ -5,13 +5,24 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/opensuse-docker-install.md]: Suggested rewrite:
so the following commands install the latest version, which includes

File: install/docker/sles-docker-install.md [hunk 6]
```diff
@@ -5,12 +5,23 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/sles-docker-install.md]: Suggested rewrite:
so the following commands install the latest version, which includes

File: install/docker/ubuntu-docker-install.md [hunk 7]
```diff
@@ -5,17 +5,28 @@
...
-
-   > **Note:** This step is only required if Docker is not installed on the system.
+1. Install Docker with the following commands:
+
+   > **Notes:**
+   >
+   > - This step is only required if Docker is not installed on the system.
+   > - Newer versions of Docker have
+   >   [removed devicemapper support](https://github.com/kata-containers/documentation/issues/373)
+   >   so the commands below install the latest version which includes
```
Comment (klynnrif) [inline install/docker/ubuntu-docker-install.md]: Suggested rewrite:
so the following commands install the latest version, which includes

**Comment by jodh-intel** [comment]
@klynnrif - thanks for reviewing. Branch updated.

@marcov - please tal (and remove the dnm label if you're happy ;)

@grahamwhaley - did you originally ack (I think we didn't save the acks way back then)? Either way, could you tal?

Ping @kata-containers/documentation - would be great to get this landed!

**Comment by grahamwhaley** [comment]
dropping the dnm. I think you will need @klynnrif to ack to clear the change request....

**Pull request merged**